core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala

```diff
@@ -1283,6 +1285,56 @@ class FsHistoryProviderSuite extends SparkFunSuite with Matchers with Logging {
     assert(deserializedOldObj.isComplete === false)
   }
 
+  test("SPARK-29755 LogInfo should be serialized/deserialized by jackson properly") {
+    def assertSerDe(serializer: KVStoreScalaSerializer, info: LogInfo): Unit = {
+      val infoAfterSerDe = serializer.deserialize(
+        serializer.serialize(info), classOf[LogInfo])
+      assert(infoAfterSerDe === info)
+      assertOptionAfterSerde(infoAfterSerDe.lastIndex, info.lastIndex)
+    }
+
+    val serializer = new KVStoreScalaSerializer()
+    val logInfoWithIndexAsNone = LogInfo("dummy", 0, LogType.EventLogs, Some("appId"),
+      Some("attemptId"), 100, None, false)
+    assertSerDe(serializer, logInfoWithIndexAsNone)
+
+    val logInfoWithIndex = LogInfo("dummy", 0, LogType.EventLogs, Some("appId"),
+      Some("attemptId"), 100, Some(3), false)
+    assertSerDe(serializer, logInfoWithIndex)
+  }
+
+  test("SPARK-29755 AttemptInfoWrapper should be serialized/deserialized by jackson properly") {
+    def assertSerDe(serializer: KVStoreScalaSerializer, attempt: AttemptInfoWrapper): Unit = {
+      val attemptAfterSerDe = serializer.deserialize(serializer.serialize(attempt),
+        classOf[AttemptInfoWrapper])
+      assert(attemptAfterSerDe.info === attempt.info)
+      // skip comparing some fields, as they've not triggered SPARK-29755
+      assertOptionAfterSerde(attemptAfterSerDe.lastIndex, attempt.lastIndex)
+    }
+
+    val serializer = new KVStoreScalaSerializer()
+    val appInfo = new ApplicationAttemptInfo(None, new Date(1), new Date(1), new Date(1),
+      10, "spark", false, "dummy")
+    val attemptInfoWithIndexAsNone = new AttemptInfoWrapper(appInfo, "dummyPath", 10, None,
+      None, None, None, None)
+    assertSerDe(serializer, attemptInfoWithIndexAsNone)
+
+    val attemptInfoWithIndex = new AttemptInfoWrapper(appInfo, "dummyPath", 10, Some(1),
+      None, None, None, None)
+    assertSerDe(serializer, attemptInfoWithIndex)
+  }
+
+  private def assertOptionAfterSerde(opt: Option[Long], expected: Option[Long]): Unit = {
+    if (expected.isEmpty) {
+      assert(opt.isEmpty)
+    } else {
+      // The issue happens only the value in Option is being unboxed. Simple comparison sometimes
+      // doesn't go though unboxing the value, hence ClassCastException is not occurred.
+      // Here we ensure unboxing to Long succeeds. Please refer SPARK-29755 for more details.
+      assert(BoxesRunTime.unboxToLong(opt.get) === expected.get)
```

Comment:
Can't you just use `.toLong` here?

Reply:
Yes.

```
scala> val i = Option(1)
i: Option[Int] = Some(1)

scala> val l = i.asInstanceOf[Option[Long]]
l: Option[Long] = Some(1)

scala> l.get.toLong
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
  at scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:105)
  ... 28 elided
```

Reply:
Here opt really is an `Option[Long]`, not a `Option[Int]`. You can't assign the latter to the former in Scala anyway. I feel like I'm missing the whole point here. Is the idea that Jackson is coming up with an `Option[Int]` when deserializing, without this change?

Reply:
Actually, that's the bug. Jackson deserializes the value "1" as an `Option[Int]` because it fits in an int, and sticks it into that variable using reflection, regardless of the type parameter (since at runtime Jackson can only see it's an `Option`, not an `Option[Long]`).

Reply:
Got it, but if the bug is fixed, then this really is an `Option[Long]`. Would handling the case that it isn't not cover up the bug if it existed? again I probably misunderstand

Reply:
I don't really understand your question.

If this test existed previously, then it would have failed. So that's the goal, right? Have a test that checks the correct behavior of something.

I'd prefer to have a test that exercised this through the normal SHS operation (e.g. trying to list applications would fail or something like that), but seems like that was not possible or not easy.

Reply:
So, before the problem is roughly that you get an `Option[Int]` where an `Option[Long]` should be. This code seems to _accommodate_ that possibility in the test, allowing it work, when we want a test that would fail if it found an `Option[Int]` here. Right? what am I missing. If it's really an `Option[Long]` as expected (post fix), then there isn't a need to accommodate an `Option[Int]`.

Reply:
I still don't understand your question.

As Jungtaek said, this test fails before the fix, because even if the code says "Option[Long]", Jackson actually put an "Option[Int]" in there.

So, as far as this test goes:
- fails without the fix
- works with the fix

Again, isn't that what we want?

Reply:
Jackson-Scala FAQ page I've linked in description explains what is happening here and why users should apply the workaround. They couldn't find a good way to fix it in Jackson-Scala side.

Why I simply pick `BoxesRunTime.unboxToLong` directly is, that's the actual path where exception is thrown. (Please refer JIRA description - https://issues.apache.org/jira/browse/SPARK-29755) 
Simple comparison (like `opt === expect`) didn't bring exception, so I wanted to make sure it definitely fails on current master, not probably fails.

> I'd prefer to have a test that exercised this through the normal SHS operation

Yeah that was what I tried first, and checkLogs() swallowed the exception (so we can't catch CCE in caller side) and another unexpected error was coming when trying to verify. It might also reflect something is not working, but looks to be less clear.

Reply:
BTW I was just commenting you can use `opt.get.toLong` to hit the bug without the fix, which looks a bit better than calling into `BoxesRunTime`, but either is fine really.

Reply:
Ah got it. No strong voice on either. I'll change it.

Reply:
Yes, `.toLong` hits the bug. So I was wondering what the deal is with `BoxesRunTime`, and though it was there to make `Option[Int]` -> `long` work. But then that would not fail the test. If that's not what it does, and it effectively does `.toLong`, then yeah we're on the same page: just use `.toLong`